diff --color -urN alloc_backup/src/alloc.rs alloc/src/alloc.rs
--- alloc_backup/src/alloc.rs	2024-09-04 20:46:24
+++ alloc/src/alloc.rs	2024-09-04 20:47:48
@@ -416,7 +416,6 @@
             panic!("memory allocation of {size} bytes failed")
         } else {
             core::panicking::panic_nounwind_fmt(
-                format_args!("memory allocation of {size} bytes failed"),
                 /* force_no_backtrace */ false,
             )
         }
